File: 41369dadf95a89691de89c2eb3d3104e5f3c1c3f/apps/api/app/schemas/blueprint.py
"""Blueprint ailesinin istek/yanıt sözleşmeleri (002 US3, T504).

Kendi modülünde yaşar: `schemas/assessment.py` bugün soru payload'ının dört tipini
ve sınav oturumu zarfını taşıyor, blueprint ise ayrı bir eksen. Aynı dosyaya
yığmak, iki şeridin aynı dosyaya dokunmasını da gerektirirdi.

**Alan adları İngilizce, kullanıcıya dönen metinler Türkçe.** Depoda kurulu ayrım
bu: `AnswerFeedbackOut.missing_points` İngilizce, içine yazılan `answers.feedback`
jsonb anahtarları (`eksik_noktalar`) Türkçe ve hata cümleleri Türkçe (Anayasa V).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.assessment import ExamVersionStatus, QuestionDifficulty, QuestionType

# ---------------------------------------------------------------------------
# Öğrenme çıktıları (FR-110)
# ---------------------------------------------------------------------------


class LearningOutcomeCreate(BaseModel):
    code: str = Field(min_length=1, max_length=40)
    description: str = Field(min_length=1, max_length=1000)
    topic_id: UUID | None = None


class LearningOutcomeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    description: str
    topic_id: UUID | None
    created_at: datetime


# ---------------------------------------------------------------------------
# Dağılım hücreleri (FR-111, FR-112)
# ---------------------------------------------------------------------------


class BlueprintCellIn(BaseModel):
    learning_outcome_id: UUID
    difficulty: QuestionDifficulty
    question_type: QuestionType
    question_count: int = Field(ge=1, le=100)
    points_per_question: int = Field(default=1, ge=1, le=100)


class BlueprintCellOut(BlueprintCellIn):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    #: Türkçe hücre adı — hata ve rapor metinleri bunu kullanır, istemci kendi
    #: etiketini uydurmaz (Anayasa V).
    label: str


class DistributionTargets(BaseModel):
    """Arayüzün aldığı MARJİNAL dağılım; hücrelerin ona uyup uymadığı sınanır.

    Yüzdeler saklanmaz: ekran "%40 kolay" alır, tam sayıya çevirir ve buraya ADET
    gönderir. Yuvarlamayı istemci yapar, ama yuvarlamanın tuttuğunu sunucu
    doğrular — SC-003'ün "birebir uyar" iddiası ancak böyle karar verilebilir olur.

    Hepsi opsiyonel: yalnız hücre göndermek de geçerli bir kullanımdır.
    """

    total_questions: int | None = Field(default=None, ge=1, le=1000)
    by_difficulty: dict[QuestionDifficulty, int] | None = None
    by_question_type: dict[QuestionType, int] | None = None


# ---------------------------------------------------------------------------
# Blueprint (FR-111)
# ---------------------------------------------------------------------------


class BlueprintCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    duration_minutes: int = Field(ge=1, le=600)
    max_attempts: int = Field(default=1, ge=1, le=100)
    opens_at: datetime | None = None
    closes_at: datetime | None = None
    cells: list[BlueprintCellIn] = Field(min_length=1, max_length=200)
    targets: DistributionTargets | None = None


class BlueprintUpdate(BaseModel):
    """Hücreler verilirse KÜME OLARAK değişir (sil + yaz).

    Tek hücrelik güncelleme bilerek yok: FR-112 doğrulaması küme üzerinde yapılır
    ve tekil bir UPDATE doğrulamayı atlayıp tutarsız bir dağılım bırakabilirdi.
    Veritabanı da aynı kararı taşıyor — `blueprint_cells`'te UPDATE ne politikası
    ne yetkisi var.
    """

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    duration_minutes: int | None = Field(default=None, ge=1, le=600)
    max_attempts: int | None = Field(default=None, ge=1, le=100)
    opens_at: datetime | None = None
    closes_at: datetime | None = None
    cells: list[BlueprintCellIn] | None = Field(default=None, min_length=1, max_length=200)
    targets: DistributionTargets | None = None


class TopicShare(BaseModel):
    """Konu dağılımı — TÜRETİLMİŞ ve salt okunur (data-model.md §8 madde 2).

    `topic_id` NULL olan çıktılar `topic_id=None` satırında toplanır ve ekranda
    "konusuz çıktıdan geliyor" diye gösterilir; sessizce yuvarlanmaz (Anayasa III).
    """

    topic_id: UUID | None
    topic_name: str | None
    question_count: int


class BlueprintOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    course_id: UUID
    title: str
    description: str | None
    duration_minutes: int
    max_attempts: int
    opens_at: datetime | None
    closes_at: datetime | None
    created_at: datetime
    updated_at: datetime
    cells: list[BlueprintCellOut]
    #: SUM(cells.question_count) — kolon değil, türetilmiş. Tutarsız olamaz.
    total_questions: int
    total_points: int
    topic_distribution: list[TopicShare]
    published_version_no: int | None = None


# ---------------------------------------------------------------------------
# Sürümler (FR-114, FR-115)
# ---------------------------------------------------------------------------


class ExamVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    blueprint_id: UUID
    version_no: int
    status: ExamVersionStatus
    published_at: datetime | None
    superseded_at: datetime | None
    created_at: datetime
    item_count: int = 0
    total_points: int = 0


class ExamItemIn(BaseModel):
    question_id: UUID
    #: Verilmezse gönderim sırası kullanılır.
    position: int | None = Field(default=None, ge=1, le=1000)


class ExamItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    position: int
    question_id: UUID
    points: int
    question_type: QuestionType
    difficulty: QuestionDifficulty | None
    learning_outcome_id: UUID | None
    stem: str


class MissingCell(BaseModel):
    """İstenen adet ile dolan adet tutmayan hücre (FR-114)."""

    learning_outcome_id: UUID
    difficulty: QuestionDifficulty
    question_type: QuestionType
    required: int
    filled: int
    label: str


class UnclassifiedItem(BaseModel):
    """Kalemde olup hiçbir hücreye sayılamayan soru (data-model.md §8 madde 7).

    Eksik hücreden AYRI raporlanır: tek listede görünseydi başka bir hücre
    eksikmiş gibi okunur, öğretmen o hücreye soru eklemeye çalışır ve gerçek sebep
    hiç söylenmemiş olurdu.
    """

    question_id: UUID
    position: int
    missing_fields: list[str]
    label: str


class ReadinessOut(BaseModel):
    """Yayın kapısının raporu. `ready` ancak İKİ liste de boşsa true olur."""

    ready: bool
    missing_cells: list[MissingCell]
    unclassified_items: list[UnclassifiedItem]
    message: str


class PublishOut(BaseModel):
    version: ExamVersionOut
    #: Yayın anında dondurulan dağılım kanıtı (data-model.md §8 madde 1).
    blueprint_snapshot: list[dict[str, Any]]
    message: str

File: 944922969a6ac2931f6dd35466da5c620386d078/apps/web/lib/blueprint.ts
/**
 * Sınav blueprint'inin saf çekirdeği: tipler, türetmeler ve yüzde→adet açılımı.
 *
 * Kendi modülünde yaşar. `lib/questions.ts` soru havuzunun payload daraltmasını
 * taşıyor, blueprint ise ayrı bir eksen; aynı dosyaya yığmak iki şeridin aynı
 * dosyaya dokunmasını da gerektirirdi.
 *
 * **Bu dosya hata METNİ üretmez.** Dağılım tutarsızsa cümleyi sunucu kurar
 * (Anayasa V: "backend tek hata zarfı üretir; frontend kendi hata metnini
 * uydurmaz"). Buradaki tek doğrulama, isteğin gönderilebilir ŞEKİLDE olup
 * olmadığıdır — "hangi hücre tutmuyor" sorusunun cevabı sunucudan gelir.
 *
 * Her fonksiyon DOM'suz koşar; `blueprint.test.ts` bunları doğrudan sınar.
 */

/* -------------------------------------------------------------------------
 * Tipler — backend `schemas/blueprint.py`'ın aynası
 * ---------------------------------------------------------------------- */

export type Difficulty = "easy" | "medium" | "hard";
export type BlueprintQuestionType = "mcq" | "open" | "code_trace" | "bug_hunt";
export type VersionStatus = "draft" | "published" | "superseded";

export interface LearningOutcome {
  id: string;
  code: string;
  description: string;
  topic_id: string | null;
  created_at: string;
}

export interface BlueprintCellInput {
  learning_outcome_id: string;
  difficulty: Difficulty;
  question_type: BlueprintQuestionType;
  question_count: number;
  points_per_question: number;
}

export interface BlueprintCell extends BlueprintCellInput {
  id: string;
  /** Türkçe hücre adı. Sunucudan gelir; ekran kendi etiketini kurmaz. */
  label: string;
}

export interface TopicShare {
  topic_id: string | null;
  topic_name: string | null;
  question_count: number;
}

export interface Blueprint {
  id: string;
  course_id: string;
  title: string;
  description: string | null;
  duration_minutes: number;
  max_attempts: number;
  opens_at: string | null;
  closes_at: string | null;
  created_at: string;
  updated_at: string;
  cells: BlueprintCell[];
  total_questions: number;
  total_points: number;
  topic_distribution: TopicShare[];
  published_version_no: number | null;
}

export interface ExamVersion {
  id: string;
  blueprint_id: string;
  version_no: number;
  status: VersionStatus;
  published_at: string | null;
  superseded_at: string | null;
  created_at: string;
  item_count: number;
  total_points: number;
}

export interface ExamItem {
  id: string;
  position: number;
  question_id: string;
  points: number;
  question_type: BlueprintQuestionType;
  difficulty: Difficulty | null;
  learning_outcome_id: string | null;
  stem: string;
}

export interface MissingCell {
  learning_outcome_id: string;
  difficulty: Difficulty;
  question_type: BlueprintQuestionType;
  required: number;
  filled: number;
  label: string;
}

export interface UnclassifiedItem {
  question_id: string;
  position: number;
  missing_fields: string[];
  label: string;
}

export interface Readiness {
  ready: boolean;
  missing_cells: MissingCell[];
  unclassified_items: UnclassifiedItem[];
  message: string;
}

/**
 * Havuz sorusunun blueprint için gereken iki ek alanı.
 *
 * `lib/types.ts`'teki `Question` genişletilmedi: o dosya bütün ekranların ortak
 * sözleşmesi ve bu turda başka bir şerit de ona dokunuyor. Blueprint'e özgü
 * alanlar burada, kendi modülünde yaşıyor (görev tanımının "blueprint frontend
 * tiplerini kendi modülünde tut" kuralı).
 */
export interface PoolQuestion {
  id: string;
  type: BlueprintQuestionType;
  payload: Record<string, unknown>;
  learning_outcome_id: string | null;
  difficulty: Difficulty | null;
}

/* -------------------------------------------------------------------------
 * Etiketler — ürün sözlüğü, tek yer
 *
 * Backend'in `DIFFICULTY_TR` / `QUESTION_TYPE_TR` sözlüğünün ekran karşılığı.
 * `labels.ts` bu şeridin dosyası değil ve reliability şeridi ona dokunuyor;
 * blueprint'e özgü etiketler burada durur.
 * ---------------------------------------------------------------------- */

export const DIFFICULTY_LABEL: Record<Difficulty, string> = {
  easy: "Kolay",
  medium: "Orta",
  hard: "Zor",
};

export const DIFFICULTIES: Difficulty[] = ["easy", "medium", "hard"];

export const VERSION_STATUS_LABEL: Record<VersionStatus, string> = {
  draft: "Taslak",
  published: "Yayında",
  superseded: "Yerine yenisi geldi",
};

/* -------------------------------------------------------------------------
 * Türetmeler
 * ---------------------------------------------------------------------- */

export function totalQuestions(cells: readonly BlueprintCellInput[]): number {
  return cells.reduce((sum, cell) => sum + cell.question_count, 0);
}

export function totalPoints(cells: readonly BlueprintCellInput[]): number {
  return cells.reduce((sum, cell) => sum + cell.question_count * cell.points_per_question, 0);
}

/** Aynı hücrenin iki kez tanımlandığını ekranda göstermek için (kayıt kapısı sunucuda). */
export function cellKey(cell: BlueprintCellInput): string {
  return `${cell.learning_outcome_id}|${cell.difficulty}|${cell.question_type}`;
}

export function hasDuplicateCells(cells: readonly BlueprintCellInput[]): boolean {
  return new Set(cells.map(cellKey)).size !== cells.length;
}

/**
 * Yüzdeyi ADETE çevirir ve toplamı TAM tutturur (en büyük artık yöntemi).
 *
 * Neden burada: "%40 kolay / %40 orta / %20 zor" 7 soruda 2,8 eder ve yuvarlama
 * kuralı saklanan veride görünmezse SC-003'ün "birebir uyar" iddiası karar
 * verilemez hâle gelir. Saklanan gerçek ADETTİR; yuvarlamayı ekran yapar, ama
 * yaptığını sunucu ayrıca doğrular (`targets`).
 *
 * Basit `Math.round` yetmez: %33/%33/%34 ile 10 soru 3+3+3=9 verir ve bir soru
 * buharlaşır. En büyük artık yöntemi toplamı her zaman `total` tutar.
 */
export function splitByShares(total: number, shares: readonly number[]): number[] {
  if (total <= 0 || shares.length === 0) return shares.map(() => 0);
  const sum = shares.reduce((acc, share) => acc + share, 0);
  if (sum <= 0) return shares.map(() => 0);

  const exact = shares.map((share) => (share / sum) * total);
  const floors = exact.map(Math.floor);
  let remaining = total - floors.reduce((acc, value) => acc + value, 0);

  const order = exact
    .map((value, index) => ({ index, remainder: value - Math.floor(value) }))
    .sort((a, b) => b.remainder - a.remainder || a.index - b.index);

  const counts = [...floors];
  for (const { index } of order) {
    if (remaining <= 0) break;
    counts[index] += 1;
    remaining -= 1;
  }
  return counts;
}

/**
 * Kapının iki listesini tek bir "ne yapmalıyım" listesine çevirir.
 *
 * İki listeyi ayrı tutmak KARARDIR (data-model.md §8 madde 7): sınıflandırılmamış
 * kalem, eksik hücre gibi okunursa öğretmen yanlış hücreye soru ekler. Ekranda da
 * ayrı başlıklar altında dururlar; bu fonksiyon yalnız sayıları verir.
 */
export function readinessCounts(readiness: Readiness | null): {
  missing: number;
  unclassified: number;
  blocked: boolean;
} {
  if (!readiness) return { missing: 0, unclassified: 0, blocked: false };
  return {
    missing: readiness.missing_cells.length,
    unclassified: readiness.unclassified_items.length,
    blocked: !readiness.ready,
  };
}

/**
 * Yayınlanmış sürümü olan bir blueprint'in hücreleri DÜZENLENEBİLİR.
 *
 * Yayınlanmış sürümün dağılım kanıtı kendi `blueprint_snapshot`'ında donmuştur;
 * düzenleme onu etkilemez. Ekran bunu söylemek zorunda, yoksa öğretmen kanıtı
 * bozduğunu sanır ve düzenlemekten çekinir.
 */
export function editingNoticeFor(blueprint: Blueprint): string | null {
  if (blueprint.published_version_no === null) return null;
  return (
    `${blueprint.published_version_no}. sürüm yayında. Dağılımı değiştirmek yayındaki ` +
    "sınavı etkilemez: o sürüm yayın anındaki dağılımı kendi içinde dondurdu. " +
    "Değişiklik ancak yeni bir sürüm yayınlarsan öğrenciye ulaşır."
  );
}
